pynbody/snapshot/gadgethdf.py

```diff
@@ -154,6 +158,302 @@ class _SubfindHdfMultiFileManager(_GadgetHdfMultiFileManager):
+            family_length = 0
+
+            ptype_slice_start = family_slice_start
+
+            for particle_type in self._family_to_group_map[fam]:
+
+                ptype_slice_len = self._load_control.mem_family_slice[particle_type].stop - self._load_control.mem_family_slice[particle_type].start
+
+                family_length += ptype_slice_len
+                ptype_slice_start += ptype_slice_len
+
+            self._family_slice_to_load[fam] = slice(family_slice_start, family_slice_start + family_length)
+            family_slice_start += family_length
+        
+
+    def load_arrays(self, all_fams_to_load: list[family.Family], sim: SimSnap, array_name: str, translated_names: list[str]):
+        """Load an array from the HDF files into the simulation snapshot.
+        
+        Parameters
+        ----------
+        all_fams_to_load : list of pynbody.family.Family
+            A list of pynbody families for which to load the array.
+        sim : SimSnap
+            The parent SimSnap object.
+        array_name : str
+            The pynbody standard name for the array to load.
+        translated_names : list of str
+            A list of possible names for the array in the Gadget HDF file.
+
+        """
+
+        array_filler = _HDFArrayFiller()
+        for loading_fam in all_fams_to_load:
+            
+            sim_fam_array = sim[loading_fam][array_name]
+            array_filler._update_sim_element_size(sim_fam_array)
+
+            i0 = 0 # current write position in sim_fam_array
+
+            # A 'gadget group name' is e.g. 'PartType0', 'PartType1' etc.
+            for hdf_group_name in self._family_to_group_map[loading_fam]:
+                if self._file_ptype_slice[hdf_group_name].stop <= self._file_ptype_slice[hdf_group_name].start:
+                    continue
+                # Create iterator for this group type across all files
+                file_iterator = _HDFFileIterator(iter(self._hdf_files.iter_particle_groups_with_name(hdf_group_name)))
+
+                last_file_index = -1
+                for readlen, buf_index, mem_index in self._load_control.iterate_with_interrupts(
+                        hdf_group_name, 
+                        hdf_group_name, 
+                        self._file_interrupt_points[hdf_group_name],  # file offset
+                        file_iterator.select_file): 
+                    if mem_index is None or file_iterator.current_hdf_file is None:
+                        continue
+                    i1 = i0 + mem_index.stop - mem_index.start
+
+                    # Check if we need to load a new dataset
+                    if last_file_index != file_iterator.file_index:
+                        dataset = self._get_dataset_from_translated_names(sim, file_iterator.current_hdf_file, translated_names)
+                        last_file_index = file_iterator.file_index
+                        array_filler._update_file_element_size(dataset)
```

Comment:
is there any sensible scenario in which different files in a series would actually present different element sizes? I am not sure why this needs updating once per file. It would be neatest if this is setup when constructing the `_HDFArrayFiller` above, then there is just a single `fill_array_from_hdf_dataset` call for each chunk